File: chat-demo-android/src/main/java/com/twilio/chat/demo/ChannelViewHolder.java
package com.twilio.chat.demo;

import com.twilio.chat.Channel;
import com.twilio.chat.Channel.ChannelStatus;
import com.twilio.chat.CallbackListener;

import android.graphics.Color;
import android.util.Log;
import android.view.View;
import android.widget.TextView;
import uk.co.ribot.easyadapter.ItemViewHolder;
import uk.co.ribot.easyadapter.PositionInfo;
import uk.co.ribot.easyadapter.annotations.LayoutId;
import uk.co.ribot.easyadapter.annotations.ViewId;

@LayoutId(R.layout.channel_item_layout)
public class ChannelViewHolder extends ItemViewHolder<Channel>
{
    @ViewId(R.id.channel_friendly_name)
    TextView friendlyName;

    @ViewId(R.id.channel_sid)
    TextView channelSid;

    @ViewId(R.id.channel_updated_date)
    TextView updatedDate;

    @ViewId(R.id.channel_created_date)
    TextView createdDate;

    @ViewId(R.id.channel_users_count)
    TextView usersCount;

    @ViewId(R.id.channel_total_messages_count)
    TextView totalMessagesCount;

    @ViewId(R.id.channel_unconsumed_messages_count)
    TextView unconsumedMessagesCount;

    View view;

    public ChannelViewHolder(View view)
    {
        super(view);
        this.view = view;
    }

    @Override
    public void onSetListeners()
    {
        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v)
            {
                OnChannelClickListener listener = getListener(OnChannelClickListener.class);
                if (listener != null) {
                    listener.onChannelClicked(getItem());
                }
            }
        });
    }

    @Override
    public void onSetValues(Channel channel, PositionInfo arg1)
    {
        friendlyName.setText(channel.getFriendlyName());
        channelSid.setText(channel.getSid());

        String updated = channel.getDateUpdatedAsDate() != null ?
                             channel.getDateUpdatedAsDate().toString() :
                             "<no updated date>";
        updatedDate.setText(updated);

        String created = channel.getDateCreatedAsDate() != null ?
                             channel.getDateCreatedAsDate().toString() :
                             "<no created date>";
        createdDate.setText(created);

        channel.getUnconsumedMessagesCount(new CallbackListener<Integer>() {
            @Override
            public void onSuccess(Integer value) {
                Log.d("ChannelViewHolder", "getUnconsumedMessagesCount callback");
                unconsumedMessagesCount.setText("Unread "+value.toString());
            }
        });

        channel.getMessagesCount(new CallbackListener<Integer>() {
            @Override
            public void onSuccess(Integer value) {
                Log.d("ChannelViewHolder", "getMessagesCount callback");
                totalMessagesCount.setText("Messages "+value.toString());
            }
        });

        channel.getMembersCount(new CallbackListener<Integer>() {
            @Override
            public void onSuccess(Integer value) {
                Log.d("ChannelViewHolder", "getMembersCount callback");
                usersCount.setText("Members "+value.toString());
            }
        });

        boolean joined = (channel.getStatus() == ChannelStatus.JOINED);
        view.setBackgroundColor(joined ? Color.WHITE : Color.GRAY);
    }

    public interface OnChannelClickListener {
        void onChannelClicked(Channel channel);
    }
}

File: chat-demo-android/src/main/java/com/twilio/chat/demo/ChannelViewHolder.kt
package com.twilio.chat.demo

import com.twilio.chat.demo.ChannelModel
import com.twilio.chat.Channel.ChannelStatus
import com.twilio.chat.Channel.ChannelType
import com.twilio.chat.CallbackListener

import android.graphics.Color
import android.util.Log
import android.view.View
import android.widget.TextView
import uk.co.ribot.easyadapter.ItemViewHolder
import uk.co.ribot.easyadapter.PositionInfo
import uk.co.ribot.easyadapter.annotations.LayoutId
import uk.co.ribot.easyadapter.annotations.ViewId

@LayoutId(R.layout.channel_item_layout)
class ChannelViewHolder(internal var view: View) : ItemViewHolder<ChannelModel>(view) {
    @ViewId(R.id.channel_friendly_name)
    internal var friendlyName: TextView? = null

    @ViewId(R.id.channel_sid)
    internal var channelSid: TextView? = null

    @ViewId(R.id.channel_updated_date)
    internal var updatedDate: TextView? = null

    @ViewId(R.id.channel_created_date)
    internal var createdDate: TextView? = null

    @ViewId(R.id.channel_users_count)
    internal var usersCount: TextView? = null

    @ViewId(R.id.channel_total_messages_count)
    internal var totalMessagesCount: TextView? = null

    @ViewId(R.id.channel_unconsumed_messages_count)
    internal var unconsumedMessagesCount: TextView? = null

    override fun onSetListeners() {
        view.setOnClickListener {
            val listener = getListener(OnChannelClickListener::class.java)
            listener?.onChannelClicked(item)
        }
    }

    override fun onSetValues(channel: ChannelModel, arg1: PositionInfo) {
        ButterKnife.bind(view)

        friendlyName.text = channel.friendlyName
        channelSid.text = channel.sid

        updatedDate.text = if (channel.dateUpdatedAsDate != null)
            channel.dateUpdatedAsDate!!.toString()
        else
            "<no updated date>"

        createdDate.text = if (channel.dateCreatedAsDate != null)
            channel.dateCreatedAsDate!!.toString()
        else
            "<no created date>"

        channel.getUnconsumedMessagesCount(object : CallbackListener<Long>() {
            override fun onSuccess(value: Long?) {
                Log.d("ChannelViewHolder", "getUnconsumedMessagesCount callback")
                unconsumedMessagesCount.text = "Unread " + value!!.toString()
            }
        })

        channel.getMessagesCount(object : CallbackListener<Long>() {
            override fun onSuccess(value: Long?) {
                Log.d("ChannelViewHolder", "getMessagesCount callback")
                totalMessagesCount.text = "Messages " + value!!.toString()
            }
        })

        channel.getMembersCount(object : CallbackListener<Long>() {
            override fun onSuccess(value: Long?) {
                Log.d("ChannelViewHolder", "getMembersCount callback")
                usersCount.text = "Members " + value!!.toString()
            }
        })

        view.setBackgroundColor(
            if (channel.status == ChannelStatus.JOINED) {
                if (channel.type == ChannelType.PRIVATE)
                    Color.BLUE
                else
                    Color.WHITE
            } else {
                if (channel.status == ChannelStatus.INVITED)
                    Color.YELLOW
                else
                    Color.GRAY
            }
        )
    }

    interface OnChannelClickListener {
        fun onChannelClicked(channel: ChannelModel)
    }
}
